File: c++/0002-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        return addTwoNumbers(l1, l2, 0);
    }

    ListNode* addTwoNumbers(ListNode *l1, ListNode *l2, int carry) {
        if(l1 == nullptr && l2 == nullptr && carry == 0) {
            return nullptr;
        }
        
        bool hasNext = false;
        int newVal = carry;

        ListNode *nl1 = nullptr;
        ListNode *nl2 = nullptr;

        if(l1 != nullptr) {
            nl1 = l1 -> next;
            newVal += l1 -> val;
        }

        if(l2 != nullptr) {
            nl2 = l2 -> next;
            newVal += l2 -> val;
        }

        hasNext = (nl1 != nullptr || nl2 != nullptr) || newVal >= 10;
        ListNode *newNode = new ListNode(newVal % 10);

        if(hasNext) {
            newNode -> next = addTwoNumbers(nl1, nl2, newVal / 10);
        }

        return newNode;
    }
};

File: dev-rust/0002-add-two-numbers.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {

    fn _add_two_numbers(l1: Option<Box<ListNode>>, l2: Option<Box<ListNode>>, carry: i32) -> Option<Box<ListNode>> {
        return match (l1, l2, carry) {
            (None, None, 0) => {
                return None;
            }

            (None, None, c) => {
                return Some(Box::new(ListNode { next: None, val: c }));
            }

            (Some(lnode), None, 0) => {
                return Some(lnode);
            }

            (Some(lnode), None, c) => {
                let new_sum   = lnode.val + c;
                let new_digit = new_sum % 10;
                let new_carry = new_sum / 10;

                let new_next = Self::_add_two_numbers(lnode.next, None, new_carry);
                let new_node = Box::new(ListNode { next: new_next, val: new_digit });
                return Some(new_node);
            }

            (None, Some(rnode), 0) => {
                return Some(rnode);
            }

            (None, Some(rnode), c) => {
                let new_sum   = rnode.val + c;
                let new_digit = new_sum % 10;
                let new_carry = new_sum / 10;

                let new_next = Self::_add_two_numbers(None, rnode.next, new_carry);
                let new_node = Box::new(ListNode { next: new_next, val: new_digit });
                return Some(new_node);
            }

            (Some(lnode), Some(rnode), c) => {
                let new_sum   = lnode.val + rnode.val + c;
                let new_digit = new_sum % 10;
                let new_carry = new_sum / 10;

                let new_next = Self::_add_two_numbers(lnode.next, rnode.next, new_carry);
                let new_node = Box::new(ListNode { next: new_next, val: new_digit });
                return Some(new_node);
            }
        }
    }

    pub fn add_two_numbers(l1: Option<Box<ListNode>>, l2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        return Self::_add_two_numbers(l1, l2, 0);
    }
}